src/sources/journald.rs

```diff
@@ -302,103 +286,138 @@ impl JournaldSource {
     async fn run_stream<'a>(
         &'a mut self,
         mut stream: BoxStream<'static, Result<Bytes, BoxedFramingError>>,
-        checkpointer: &'a mut Checkpointer,
-        cursor: &'a mut Option<String>,
+        checkpointer: &'a mut StatefulCheckpointer,
     ) -> bool {
+        let batch_size = self.batch_size;
         loop {
-            let mut record_size = 0;
-            let mut count = 0;
-            let mut byte_size = 0;
-            let mut events = Vec::new();
-            let (batch, receiver) = BatchNotifier::maybe_new_with_receiver(self.acknowledgements);
-            let mut exiting = None;
-
-            for _ in 0..self.batch_size {
-                let bytes = match stream.next().await {
-                    None => {
-                        warn!("Journalctl process stopped.");
-                        exiting = Some(true);
-                        break;
-                    }
-                    Some(Ok(text)) => text,
-                    Some(Err(error)) => {
-                        error!(
-                            message = "Could not read from journald source.",
-                            %error,
-                        );
+            let mut batch = Batch::new(self, checkpointer);
+
+            if !batch.handle_next(stream.next().await) {
+                break true;
+            }
+
+            let timeout = tokio::time::sleep(BATCH_TIMEOUT);
+            tokio::pin!(timeout);
```

Comment:
I suppose an alternate approach might be to have an [`Interval`](https://docs.rs/tokio/latest/tokio/time/fn.interval.html) in the struct itself and poll that for ticks, although it wouldn't be lined up neatly with batches like this is.

Reply:
I think that would be unwelcome for sites where journald doesn't actually emit a lot of logs, and unnecessary for sites where it does.